common/tools/dev-tool/src/util/customization/imports.ts

```diff
@@ -1,44 +1,112 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license
 
-import { Identifier, ImportDeclaration, ImportSpecifier, SourceFile, SyntaxKind } from "ts-morph";
+import { ImportDeclaration, SourceFile } from "ts-morph";
 import { getCustomizationState } from "./state";
-import * as path from "path";
+import * as path from "upath";
+
+type RelativePath = string & { __relativePath: never };
+type DotPrefixedRelativePath = string & { __dotPrefixedRelativePath: never } & RelativePath;
+type LocalModuleSpecifier = string & { __localModuleSpecifier: never } & RelativePath;
```

Comment:
Many of the functions in these files have unclear preconditions/postconditions. When these are unsatisfied, the resultant bugs are often subtle and difficult to detect. I'm concerned that, without these types, it'd be too easy to modify this file without realizing that you've failed to establish a precondition in a distant part of the file. This problem will only be exacerbated as we refactor this to support exports as well. I'm all ears for suggestions though.

Maybe this can be cleaned up by reimplementing these types as classes?

Reply:
I don't think that runtime checks for these types would be feasible. There are multiple functions that have undefined behavior when their parameter doesn't satisfy all of:

* is a string,
* is a module specifier (i.e. is or is derived from `ImportDeclaration.getModuleSpecifierValue() : string` called on an import declaration that came from a error-free TS file's corresponding `SourceFile`),
* is a local module specifier (i.e. a relative path from a source file to a module, which isn't in `node_modules`),
* is a POSIX-style path,
* and is prefixed with `./` or `../`

Some of these can be checked with a type guard. However, as there's multiple nested functions with this precondition, we'd need to call these type guards on the same immutable data several times. Plus, returning undefined or throwing an error due to something that can be guaranteed at compile time would obscure the happy path.

Some of these preconditions still can only really be asserted by keeping track of where the data came from. This could be accomplished by putting the data in an object with a set of boolean flags to track this metadata, but that's a lot of overhead and complexity that the compiler spares us in this case.

It's a bit arcane and intimidating, so I've added some comments to help others understand the syntax when they onboard. Like we talked about offline, I agree that this is a good place to experiment with this pattern and see if it brings value moving forward.